src/descriptors/keys.rs

```diff
@@ -0,0 +1,137 @@
+use miniscript::{
+    bitcoin::{
+        self,
+        hashes::{hash160, ripemd160, sha256},
+        util::bip32,
+    },
+    hash256, MiniscriptKey, ToPublicKey,
+};
+
+use std::{error, fmt, str};
+
+#[derive(Debug)]
+pub enum DescKeyError {
+    DerivedKeyParsing,
+}
+
+impl std::fmt::Display for DescKeyError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> std::fmt::Result {
+        match self {
+            DescKeyError::DerivedKeyParsing => write!(f, "Parsing derived key"),
+        }
+    }
+}
+
+impl error::Error for DescKeyError {}
+
+/// A public key used in derived descriptors
+#[derive(Debug, Eq, PartialEq, Clone, Ord, PartialOrd, Hash)]
+pub struct DerivedPublicKey {
```

Comment:
Maybe later we can user https://docs.rs/miniscript/latest/miniscript/descriptor/struct.SinglePub.html
but for now I am not sure it is implementing enough traits like fmt::Display

Reply:
Good call. Another advantage of our own `DerivedPublicKey` is that the type enforces an origin be always present.